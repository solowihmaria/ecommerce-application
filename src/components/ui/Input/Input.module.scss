.inputContainer {
    display: flex;
    flex-direction: column;
    gap: 0.25rem;
    width: 100%;
    max-width: 16rem;
}

.inputWrapper {
    position: relative;
    width: 100%;
}

.input {
    box-sizing: border-box;
    width: 100%;
    padding: 0.75rem 2.5rem 0.75rem 1rem;
    border: 1px solid #ddd;
    border-radius: 0.25rem;
    font-family: inherit;
    font-size: 1rem;
    transition: border-color 0.2s ease;

    &:focus {
        border-color: #555;
        outline: none;
    }

    &:disabled {
        cursor: not-allowed;
        background-color: #f5f5f5;
    }

    &.error {
        border-color: #d32f2f;
    }
}

.toggleButton {
    cursor: pointer;
    position: absolute;
    top: 55%;
    right: 0.75rem;
    transform: translateY(-50%);
    padding: 0;
    border: none;
    color: #666;
    background: none;

    &:hover {
        color: #333;
    }

    &:focus {
        outline: none;
    }
}

.errorMessage {
    display: block;
    width: 100%;
    min-height: 1.25rem;
    font-size: 0.7rem;
    color: #d32f2f;
    overflow-wrap: break-word;
    white-space: normal;
}
