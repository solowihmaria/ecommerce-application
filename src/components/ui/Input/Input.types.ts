import type { InputHTMLAttributes } from 'react';

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
    error?: boolean;
    errorMessage?: string;
}

export interface PasswordInputProps
    extends InputHTMLAttributes<HTMLInputElement> {
    showPassword: boolean;
    onTogglePassword: () => void;
    error?: boolean;
    errorMessage?: string;
}
