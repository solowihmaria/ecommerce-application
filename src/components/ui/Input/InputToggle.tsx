import React from 'react';
import { FaEye, FaEyeSlash } from 'react-icons/fa';
import styles from './Input.module.scss';
import type { InputToggleButtonProps } from './Input.types';

export const InputToggleButton = ({
    showPassword,
    onTogglePassword,
}: InputToggleButtonProps): React.JSX.Element => (
    <button
        type="button"
        className={styles.toggleButton}
        onClick={onTogglePassword}
        aria-label={showPassword ? 'Hide password' : 'Show password'}
    >
        {showPassword ? <FaEye /> : <FaEyeSlash />}
    </button>
);
