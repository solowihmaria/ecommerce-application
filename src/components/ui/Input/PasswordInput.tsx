import React from 'react';
import { FaEye, FaEyeSlash } from 'react-icons/fa';
import styles from './Input.module.scss';
import clsx from 'clsx';
import { FiAlertCircle } from 'react-icons/fi';
import type { PasswordInputProps } from './Input.types';

export const PasswordInput = React.forwardRef<
    HTMLInputElement,
    PasswordInputProps
>(
    (
        {
            className = '',
            error,
            errorMessage,
            isPasswordVisible,
            onVisibilityToggle,
            ...props
        },
        reference
    ) => {
        return (
            <div className={styles.inputContainer}>
                <div className={styles.inputWrapper}>
                    <input
                        ref={reference}
                        type={isPasswordVisible ? 'text' : 'password'}
                        className={clsx(
                            styles.input,
                            styles.password,
                            { [styles.error]: error },
                            className
                        )}
                        {...props}
                    />
                    <button
                        type="button"
                        className={styles.toggleButton}
                        onClick={onVisibilityToggle}
                        aria-label={
                            isPasswordVisible
                                ? 'Hide password'
                                : 'Show password'
                        }
                    >
                        {isPasswordVisible ? <FaEyeSlash /> : <FaEye />}
                    </button>
                </div>
                {error && errorMessage && (
                    <div className={styles.errorMessage}>
                        <FiAlertCircle className={styles.errorIcon} />
                        <span data-testid="error-input-password-test-id">
                            {errorMessage}
                        </span>
                    </div>
                )}
            </div>
        );
    }
);

PasswordInput.displayName = 'PasswordInput';
