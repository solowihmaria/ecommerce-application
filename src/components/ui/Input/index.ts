export { Input } from './Input';
export type { InputProps } from './Input.types';
export { PasswordInput } from './PasswordInput';
