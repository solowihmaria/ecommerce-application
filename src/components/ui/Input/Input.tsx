import React from 'react';
import styles from './Input.module.scss';
import type { InputProps } from './Input.types';
import { InputToggleButton } from './InputToggle';
import clsx from 'clsx';

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
    (
        {
            type = 'text',
            className = '',
            error = false,
            errorMessage,
            hasPasswordToggle = false,
            onTogglePassword,
            showPassword,
            ...props
        },
        reference
    ) => (
        <div
            data-testid="input-test-id"
            className={clsx(
                styles.inputContainer,
                type === 'checkbox' && styles.checkbox
            )}
        >
            <div className={styles.inputWrapper}>
                <input
                    ref={reference}
                    type={showPassword ? 'text' : type}
                    className={clsx(
                        styles.input,
                        error && styles.error,
                        hasPasswordToggle && styles.password,
                        className
                    )}
                    {...props}
                />
                {hasPasswordToggle &&
                    type === 'password' &&
                    onTogglePassword && (
                        <InputToggleButton
                            showPassword={!!showPassword}
                            onTogglePassword={onTogglePassword}
                        />
                    )}
            </div>
            {error && (
                <span className={styles.errorMessage}>{errorMessage}</span>
            )}
        </div>
    )
);

Input.displayName = 'Input';
