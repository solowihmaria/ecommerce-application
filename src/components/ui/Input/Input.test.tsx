import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Input } from './Input';

describe('<Input/> component', () => {
    test('Render Input component - without any properties', () => {
        render(<Input />);
        const inputElement = screen.getByTestId('test-id-input-field');
        expect(inputElement).toBeInTheDocument();
        expect(inputElement).toMatchSnapshot();
    });

    test('Render Input component - disabled input', () => {
        render(<Input placeholder="input test" disabled={true} />);
        const input = screen.getByPlaceholderText(/input test/i);
        expect(input).toBeInTheDocument();
        expect(input).toBeDisabled();
        expect(input).toMatchSnapshot();
    });

    test('Render Input component - function is called onChange', () => {
        const onChange = jest.fn();
        render(<Input placeholder="input test" onChange={onChange} />);
        const input = screen.getByPlaceholderText(/input test/i);
        fireEvent.change(input, { target: { value: 'Some text' } });
        expect(input).toBeInTheDocument();
        expect(onChange).toHaveBeenCalledTimes(1);
    });

    test('Render Input component - check error message', () => {
        const testErrorMessage = 'input error message';
        render(
            <Input
                placeholder="input test"
                error={true}
                errorMessage={testErrorMessage}
            />
        );
        const inputElement = screen.getByTestId('test-id-input-field');
        expect(inputElement).toBeInTheDocument();
        expect(inputElement).toMatchSnapshot();
        const errorElement = screen.getByText(testErrorMessage);
        expect(errorElement).toBeInTheDocument();
    });

    test('Render Input component - function is called onBlur', () => {
        const onBlur = jest.fn();
        render(<Input placeholder="input test" onBlur={onBlur} />);
        const input = screen.getByPlaceholderText(/input test/i);
        fireEvent.blur(input);
        expect(input).toBeInTheDocument();
        expect(onBlur).toHaveBeenCalledTimes(1);
    });
});
