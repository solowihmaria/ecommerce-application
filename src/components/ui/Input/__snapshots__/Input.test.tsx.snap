// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> component Render Input component - check error message 1`] = `
<input
  class="input error"
  data-testid="test-id-input-field"
  placeholder="input test"
  type="text"
/>
`;

exports[`<Input/> component Render Input component - disabled input 1`] = `
<input
  class="input"
  data-testid="test-id-input-field"
  disabled=""
  placeholder="input test"
  type="text"
/>
`;

exports[`<Input/> component Render Input component - without any properties 1`] = `
<input
  class="input"
  data-testid="test-id-input-field"
  type="text"
/>
`;
